week9/answers.sql

```diff
@@ -1,3 +1,148 @@
 -- Create your tables, views, functions and procedures here!
 CREATE SCHEMA social;
 USE social;
+
+
+
+CREATE TABLE users(
+user_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+first_name VARCHAR(50) NOT NULL,
+last_name VARCHAR(50) NOT NULL,
+email VARCHAR(100) NOT NULL,
+created_on TIMESTAMP NOT NULL DEFAULT NOW()
+);
+
+CREATE TABLE sessions(
+session_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+user_id INT UNSIGNED,
+created_on TIMESTAMP NOT NULL DEFAULT NOW(),
+updated_on TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+FOREIGN KEY (user_id) REFERENCES users(user_id) 
+    ON DELETE CASCADE
+    ON UPDATE CASCADE
+);
+
+CREATE TABLE friends(
+user_friend_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+user_id INT UNSIGNED,
+friend_id INT UNSIGNED,
+FOREIGN KEY (user_id) REFERENCES users(user_id) 
+    ON DELETE CASCADE
+    ON UPDATE CASCADE,
+FOREIGN KEY (friend_id) REFERENCES users(user_id) 
+    ON DELETE CASCADE
+    ON UPDATE CASCADE
+);
+
+
+CREATE TABLE posts (
+    post_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+    user_id INT UNSIGNED NOT NULL,
+    content VARCHAR(250) NOT NULL,
+    created_on TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_on TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+    FOREIGN KEY (user_id) REFERENCES users(user_id) 
+        ON DELETE CASCADE
+        ON UPDATE CASCADE
+);
+
+CREATE TABLE notifications(
+notification_id INT UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+user_id INT UNSIGNED NOT NULL,
+post_id INT UNSIGNED NOT NULL,
+FOREIGN KEY (user_id) REFERENCES users(user_id)
+    ON DELETE CASCADE
+    ON UPDATE CASCADE,
+FOREIGN KEY (post_id) REFERENCES posts(post_id) 
+    ON DELETE CASCADE
+    ON UPDATE CASCADE
+);
+
+
+
+-- notification view 
+CREATE VIEW notification_posts AS
+SELECT
+    notif.user_id AS notification_user_id,
+    u.first_name,
+    u.last_name,
+    p.post_id,
+    p.content
+FROM 
+    notifications notif
+INNER JOIN posts p ON notif.post_id = p.post_id
+INNER JOIN users u ON p.user_id = u.user_id;
+
+
+
+
+DELIMITER $$
+
+CREATE PROCEDURE AddNewUser(
+    IN new_first_name VARCHAR(50),
+    IN new_last_name VARCHAR(50),
+    IN new_email VARCHAR(100)
+)
+BEGIN
+    DECLARE new_user_id INT;
+
+    -- Insert the new user into the users table
+    INSERT INTO users (first_name, last_name, email)
+    VALUES (new_first_name, new_last_name, new_email);
+
+    -- Get the new user's ID
+    SET new_user_id = LAST_INSERT_ID();
+
+    -- Insert notifications for all existing users except the new user
+    INSERT INTO notifications (user_id, post_id)
```

Comment:
Note: This is also intended to be a CURSOR assignment (not using nested SELECT statements).